File: client_flutter/lib/classes/data_package.dart
import 'package:betalk/classes/message.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:uuid/uuid.dart';

part 'data_package.g.dart';

@JsonSerializable()
class DataPackage {
  const DataPackage(
    this.id, {
    required this.type,
    this.loginName,
    this.message,
  })  : assert(message != null || loginName != null),
        assert(!(type == DataPackageType.message && message == null)),
        assert(!(type == DataPackageType.login && loginName == null));

  DataPackage.newMessagePackage(this.message)
      : id = const Uuid().v4(),
        type = DataPackageType.message,
        loginName = null;

  DataPackage.newLoginPackage(this.loginName)
      : id = const Uuid().v4(),
        type = DataPackageType.login,
        message = null;

  final String id;
  final String? loginName;
  final Message? message;
  final DataPackageType type;

  factory DataPackage.fromJson(Map<String, dynamic> json) =>
      _$DataPackageFromJson(json);
  Map<String, dynamic> toJson() => _$DataPackageToJson(this);
}

enum DataPackageType {
  @JsonValue("login")
  login,

  @JsonValue("message")
  message,
}

File: server_kotlin/src/main/kotlin/classes/DataPackage.kt
package com.perceivers25.betalk.classes

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.*

/**
 * Data package is a class that defines a data package in the Betalk Protocol able to represent a login or a message.
 * At anytime message or login name should be specified.
 *
 * @param id The universally unique identifier of the data package
 * @param message The message itself if  type is message
 * @param loginName The user login if type is login
 * @see Server
 */
@Serializable
class DataPackage(val id: String, val type: DataPackageType, val message: Message?, val loginName: String?) {

    init {
        assert(message != null || loginName != null)
        assert(!(type == DataPackageType.Message && message == null))
        assert(!(type == DataPackageType.Login && loginName == null))
    }

    fun toJson(): String {
        val format = Json { encodeDefaults = true }
        return format.encodeToString(this);
    }

    companion object {
        fun fromJson(jsonString: String): DataPackage {
            return Json.decodeFromString<DataPackage>(jsonString)
        }

        fun newMessageDataPackage(message: Message): DataPackage {
            return DataPackage(
                UUID.randomUUID().toString(), DataPackageType.Message, message, null,
            )
        }
    }

    @Serializable
    enum class DataPackageType {
        @SerialName("login")
        Login,

        @SerialName("message")
        Message,
    }
}